### ISSUE TITLE
Specifying git commit doesn't work if plugin is already installed

### ISSUE THREAD

**Issue by akinsho** [wbthomason/packer.nvim#211]
Off the back of https://github.com/wbthomason/packer.nvim/issues/200#issuecomment-777848523 I just discovered and used the commit key which is a really nice feature, my initial attempt involved just adding a specific commit to an already installed plugin which resulted in an invalid commit error, specifying the full commit not just the first characters resulted in an even more obscure git tree error. Finally deleting the plugin and re-installing it with the key specified did work.

Ideally just adding a commit to an already installed plugin as a key would be enough

PS: apologies think I'm averaging an issue a day at this rate 😓

**Comment by wbthomason** [comment]
Thanks for the report; sorry for all the bugs 😬

I'll admit that much of the git commands used came from reading the source of other plugin managers, and this feature has gone largely unused.

This case is supposed to be handled already; `packer` should pull missing commits and then check out the right commit. 

From what you're describing, it sounds like there's a problem with fetching missing commits.

**Comment by cwebster2** [comment]
The issue is when you install normally, it does a shallow clone of depth one, which means the only commit you have is whatever HEAD was when you installed it.  Thus, if you add a commit key after the fact, when packer, (or the git cli) attempts to check out that ref, it doesn't exist in your clone and fails.

**Comment by wbthomason** [comment]
Yep! The thing is, `packer` is already supposed to handle this deepening as needed, so there's some bug in either detecting that we need to fetch additional commits or in performing the actual fetching.

**Comment by akinsho** [comment]
Just wanted to add here whilst it's still fresh in my mind that at leas on my mac trying to remove the `commit` key so a plugin is no longer pinned to a commit causes packer to hang. In my case I was removing a commit key from `gitsigns` and `:PackerSync` seemed to stall till I gave up. I just commented out the plugin removed it and re-installed without the key and that worked fine

**Comment by miguelcaravantes** [comment]
i have the same issue, i was wondering why specify the commit doesn't working, now i know the workaround is uninstall the package first, thank you
